File: Leetcode/Algorithm/java/Largest Rectangle in Histogram.java
public class Solution {
    public int largestRectangleArea(int[] height) {
        Stack<Integer> st = new Stack<Integer>();
        int n = height.length;
        int ans = 0;
        for (int i = 0; i <= n; i++) {
            int now = i < n? height[i]: 0;
            if (st.empty() || now > height[st.peek()]) {
                st.push(i);
                continue;
            }
            while (!st.empty() && now <= height[st.peek()]) {
                int p = st.pop();
                int left = st.empty()? 0: st.peek() + 1;
                int right = i - 1 ;
                ans = Math.max(ans, height[p] * (right - left + 1));
            }
            st.push(i);
        }
        return ans;
    }
}

File: LintCode/cpp/Largest Rectangle in Histogram.cc
class Solution {
public:
    /**
     * @param height: A list of integer
     * @return: The area of largest rectangle in the histogram
     */
    int largestRectangleArea(vector<int> &height) {
        height.push_back(-1);
        int n = height.size();
        stack<int> st;
        
        int ans = 0;
        for (int i = 0; i < n; i++) {
            int u = height[i];
            
            while (!st.empty() && height[st.top()] > u) {
                int v = st.top();
                st.pop();
                int l = st.empty()? 0: st.top() + 1;
                int r = i - 1;
                ans = max(ans, (r - l + 1) * height[v]);
            }
            
            st.push(i);
        }
        return ans;
    }
};

